### ISSUE TITLE
Combining Server s

### ISSUE THREAD

**Issue by barischrooneyj** [haskell-servant/servant#1085]
Hi, I am writing a library which has a function that returns a server given some configuration information. `server :: Config -> Server GlobalAPI`. In another application I want to combine this somewhat generic `Server` with some additional endpoints, in particular for static files. Is there some way of writing something like `Server :<|> Server` ?

**Comment by alpmestan** [comment]
Yes. In fact, you can write exactly what you wrote, it will be a valid server for the API defined by `GlobalAPI :<|> AdditionalAPI`.

In general, if you have `server1 :: Server api1` and `server2 :: Server api2`, then `server1 :<|> server2 :: Server (api1 :<|> api2)`. And if you use a cusom monad, it works just the same: if you have `server1 :: ServerT api1 m` and `server2 :: ServerT api2 m`, then `server1 :<|> server2 :: ServerT (api1 :<|> api2) m`.

This all works regardless of whether `api1` and `api2` are single endpoints or entire sections of your final API. Be aware though that when you defined endpoints in sub APIs, in a nested way like this, this gets reflected on the `Server` type, see the `FactoringAPI` example at https://haskell-servant.readthedocs.io/en/latest/cookbook/structuring-apis/StructuringApis.html.

**Comment by saurabhnanda** [comment]
@alpmestan is there any way of combining servers that are in different custom monad, i.e.:

```
ServerT api1 m1 :<|> ServerT api1 m2
```

**Comment by alpmestan** [comment]
I guess you can, but what you have in the end isn't a `ServerT (api1 :<|> api2) <some monad>`, just two groups of "incompatible" handlers. By the time you want to do anything uniform over all of them (e.g hoisting, serving), you _do_ want them to run in a common monad, may it be `m1`, `m2`, `Handler` or another one. Because the type signatures of hoisting/serving functions require you to do so, then only accept `ServerT api m`-shaped values, which can't ever be unified with a non-uniform bunch of handlers like in your example.

**Comment by saurabhnanda** [comment]
@alpmestan thanks for your reply. Would you recommended combining such applications at the Wai/Middleware level intead? Is there something in the cookbook about this?

**Comment by erewok** [comment]
While not a direct answer to your question, you may find some useful details in two Matt Parsons blogposts, one on [using servant as a reverse proxy for another site](https://www.parsonsmatt.org/2016/06/24/take_over_an_api_with_servant.html) implemented in another language (Ruby) and another one on [mounting a Servant app as a Yesod subsite](https://www.parsonsmatt.org/2016/12/18/servant_in_yesod_-_yo_dawg.html) inside a Yesod app.

**Comment by alpmestan** [comment]
@saurabhnanda The answer depends on what you get out of each solution and how beneficial that is. For example, in https://github.com/alpmestan/servant-prefix, I show a little trick to "mount" a servant API under a given string prefix; this has the benefit that if you're using the "prefixed API", you can even derive correct links for it, etc. Now if there's no interest in a benefit like this, I'd probably use the `Raw` approach or indeed a tiny wai middleware that dispatches on the right app based on whatever bits of the request are relevant. Or run two separate servers and dispatch at the nginx level. But frankly I personally haven't had much of an occasion to face a problem where I had to accomplish something like this, so this is all based on what I saw other people do.

I do not think there's a recipe about this, but I'm sure it would be helpful. Matt's blog posts are good references in the meantime.

**Comment by saurabhnanda** [comment]
I'm trying the simplest approach by converting one of the servant apps into a `Wai.Application` and combining it with the other servant app using `Raw`. However, I am unable to make the following type-check:

```haskell
type CombinedRoutes = (ToServant Routes1 AsApi) :<|> Routes2
data Routes1 route = Routes1 { _whatever }
data Routes2 = "prefix" :> Raw

server :: ServerT CombinedRoutes AppM
server = 
  (toServant routes1Server) :<|> waiApplication 

routes1Server :: Routes1 (AsServerT AppM)
routes1Server = _whatever

waiApplication :: Wai.Application
waiApplication = genericServe routes3Server

routes3Server :: Routes3 AsServer
routes3Server = _whatever

data Routes3 route = Routes3  { _whatever }
```

Here's the compile error that I get in my actual code:

```
/Users/saurabhnanda/projects/vl-gitlab/haskell/src/AppServer.hs:263:3: error:
    • Couldn't match type ‘Request
                           -> (Response -> IO ResponseReceived) -> IO ResponseReceived’
                     with ‘Tagged (ReaderT Env IO) Application’
      Expected type: ServerT Routes AppM
        Actual type:  -- entire API type comes here
```

**Comment by saurabhnanda** [comment]
In my previous example,`waiApplication` is basically in `IO`, whereas rest of the app is in `AppM`. Is this what is causing the compiler error?

**Comment by alpmestan** [comment]
You just need to wrap the `Application` with a `Tagged` constructor I think.
